pkg/server/router/did.go

```diff
@@ -263,21 +279,83 @@ func (dr DIDRouter) ListDIDsByMethod(c *gin.Context) {
 			return
 		}
 	}
-
 	// TODO(gabe) check if the method is supported, to tell whether this is a bad req or internal error
 	// TODO(gabe) differentiate between internal errors and not found DIDs
 	getDIDsRequest := did.ListDIDsRequest{Method: didsdk.Method(*method), Deleted: getIsDeleted}
-	gotDIDs, err := dr.service.ListDIDsByMethod(c, getDIDsRequest)
+
+	pageSizeStr := framework.GetParam(c, PageSizeParam)
+
+	if pageSizeStr != nil {
+		pageSize, err := strconv.Atoi(*pageSizeStr)
+		if err != nil {
+			errMsg := fmt.Sprintf("list DIDs by method request encountered a problem with the %q query param", PageSizeParam)
+			framework.LoggingRespondErrMsg(c, errMsg, http.StatusBadRequest)
+			return
+		}
+		getDIDsRequest.PageSize = &pageSize
+	}
+
+	queryPageToken := framework.GetParam(c, PageTokenParam)
+	if queryPageToken != nil {
+		errMsg := "token value cannot be decoded"
+		tokenData, err := base64.RawURLEncoding.DecodeString(*queryPageToken)
+		if err != nil {
+			framework.LoggingRespondErrMsg(c, errMsg, http.StatusBadRequest)
+			return
+		}
+		var pageToken PageToken
+		if err := json.Unmarshal(tokenData, &pageToken); err != nil {
+			framework.LoggingRespondErrMsg(c, errMsg, http.StatusBadRequest)
+			return
+		}
+		pageTokenValues, err := url.ParseQuery(pageToken.EncodedQuery)
+		if err != nil {
+			framework.LoggingRespondErrMsg(c, errMsg, http.StatusBadRequest)
+			return
+		}
+
+		query := pageTokenQuery(c)
+		if !reflect.DeepEqual(pageTokenValues, query) {
+			logrus.Warnf("expected query from token to be equal to query from request. token: %v\nrequest%v", pageTokenValues, query)
+			framework.LoggingRespondErrMsg(c, "page token must be for the same query", http.StatusBadRequest)
+			return
+		}
+		getDIDsRequest.PageToken = &pageToken.NextPageToken
+	}
+
+	listResp, err := dr.service.ListDIDsByMethod(c, getDIDsRequest)
 	if err != nil {
 		errMsg := fmt.Sprintf("could not get DIDs for method: %s", *method)
 		framework.LoggingRespondErrWithMsg(c, err, errMsg, http.StatusInternalServerError)
 		return
 	}
 
-	resp := ListDIDsByMethodResponse{DIDs: gotDIDs.DIDs}
+	resp := ListDIDsByMethodResponse{
+		DIDs: listResp.DIDs,
+	}
+	if listResp.NextPageToken != "" {
+		tokenQuery := pageTokenQuery(c)
+		pageToken := PageToken{
+			EncodedQuery:  tokenQuery.Encode(),
```

Comment:
Can people 'hack this' and encode other random query data to get like 'the wrong page' lol